[62525] Non-working days can be choosen via date field (2)

HDinger opened the pull request:
# Ticket

https://community.openproject.org/projects/openproject/work_packages/62525/activity


# What are you trying to accomplish?
This is an alternative approach to https://github.com/opf/openproject/pull/18611 which does not exclude active Element from the morhping to show when the SetAttributesService silently updated the date to the next non-working day

HDinger committed:
Do not exclude active Element from the morhping to show when the SetAttributesService silently updated the date to the next non-working day

cbliard commented on spec/features/work_packages/datepicker/datepicker_non_working_day_spec.rb:
```diff
@@ -45,11 +45,13 @@
     create(:non_working_day,
            date: Time.zone.today.end_of_year.next_occurring(:tuesday))
   end
+  shared_let(:working_day_next_year) { Time.zone.today.end_of_year.next_occurring(:wednesday) }
```
Test will fail if year starts with a Wednesday
2031-01-01 is a Wednesday. I tried setting my clock to 2030 to see if the test fails and it fails: `expected to find visible field "work_package[due_date]" that is not disabled with value "2031-01-01" within #<Capybara::Node::Element tag="div" path="//HTML[1]/BODY[1]/OPCE-MODAL-OVERLAY[1]/DIV[2]/NG-COMPONENT[1]/DIV[1]/TURBO-FRAME[1]/DIV[1]"> but there were no matches. Also found "", which matched the selector but not all filters.` **`Expected value to be Wed, 01 Jan 2031 but was "2031-01-08"`**

I'll fix it.

cbliard replied:
Fixed

cbliard reviewed (commented):
I really like this new approach, especially as it's much simpler. I thought there would be some issues with focus or cursor position but Idiomorph seems to handle it nicely.

Sadly it's not suitable for progress modal fields because the unit would be appended after the value every time and that's not usable:

[fighting the edit field.webm](https://github.com/user-attachments/assets/44a8bcb4-3462-4a33-a31e-57ac97105a6c)

I'll fix it.

cbliard committed:
Make sure the test won't fail in 2030

Because in 2031-01-01 would be a Wednesday, and then the expected dates would be wrong.

cbliard committed:
Do not morph active element value in progress preview

Morphing the active element value made the progress modal unusable as
the unit would be added to the active input field while typing a number
or deleting characters.

cbliard committed:
Fix flaky spec

Spec was ./spec/features/work_packages/scheduling/scheduling_mode_spec.rb:139
on job run https://github.com/opf/openproject/actions/runs/14519804438/job/40741624438

In `DateEditField#set_value`, the input fields of the date picker were
cleared and then the values were set. Problem is that when clearining
fields it does not wait for the preview to be applied: when the preview
comes back, it morphs the fields to set empty values but the
`#set_value` method has already started to fill the values, so the start
date field is overwritten with an empty value. Then the test fails
because the start date field is not the one expected.

The solution is to wait for the changes to be applied after clearing the
fields.

I wonder if that may be hiding a real bug where user could have their
input be reset while typing.

cbliard committed:
Fix flaky spec

Failing spec ./spec/features/activities/work_package/activities_spec.rb:539
on job run https://github.com/opf/openproject/actions/runs/14531919190/job/40773106633

The test did not correctly wait for the second journals changes to be
displayed. It was waiting with
`activity_tab.expect_journal_changed_attribute(text: "Subject")` but as
the subject was already changed in the first journal entry, it was
returning immediately and did not wait for the second journal entry to
be updated.

Waiting for `activity_tab.expect_journal_changed_attribute(text: "A new
subject")` instead forces the test to wait for the second journal entry
to be displayed and fixed the flakyness.
